### PR TITLE
prov/gni: Open MR notifier per domain, add locking

### PR DESCRIPTION
**Pull request by ztiffany**
kdreg limits us to opening a single MR notifier object per process.  Currently,
the provider uses a new notifier object per fabric.  To allow multiple fabrics
in a single process, create a single, global kdreg MR notifier and share
between all domains on all fabrics.  Use of multiple domains is still limited
by issue #1064.

Fixes ofi-cray/libfabric-cray#1069.

Signed-off-by: Zach <ztiffany@cray.com>

@sungeunchoi @jswaro @hppritcha

### CODE REVIEW

**Comment by jswaro** [comment]
Instead of doing refs via int, why not use the gnix reference objects?

File: prov/gni/src/gnix_mr_notifier.c [hunk 1]
```diff
@@ -55,69 +57,69 @@ notifier_verify_stuff(struct gnix_mr_notifier *mrn) {
...
-		GNIX_INFO(FI_LOG_MR, "mr notifier NULL\n");
-		return -FI_EINVAL;
-	}
-
-	mrn->fd = 0;
-	mrn->cntr = NULL;
-	fastlock_init(&mrn->lock);
+	global_mr_not.fd = -1;
+	global_mr_not.cntr = NULL;
+	fastlock_init(&global_mr_not.lock);
```
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c:64]: Shouldn't the reference count be set here?

Comment (jswaro) [reply prov/gni/src/gnix_mr_notifier.c:64]: Additionally, how does this mod address the possibility of multiple fabrics? The next fabric to be opened using the GNI provider would overwrite an already initialized notifier.

File: prov/gni/src/gnix_mr_notifier.c [hunk 2]
```diff
@@ -55,69 +57,69 @@ notifier_verify_stuff(struct gnix_mr_notifier *mrn) {
...
         kdreg_get_user_delta_args_t get_user_delta_args;
 
-	if ((mrn->fd != 0) || (mrn->cntr != NULL)) {
-		GNIX_INFO(FI_LOG_MR, "mr notifier already open\n");
-		return -FI_EBUSY;
-	}
+	fastlock_acquire(&global_mr_not.lock);
 
-	fastlock_acquire(&mrn->lock);
+	if (global_mr_not.fd < 0) {
```
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c]: A single 'if' containing the already initialized code would remove all of the whitespace changes.

```C 
if (global_mr_not.fd >= 0)
   // skip all of this and return.
```

Additionally, it would simply the exit case if you did the following:

```C

if (not-initialized)
    goto done;
...
     //code for normal open
done:
    increment reference++
```
If the reference count was initialized to 0 in the init function, then ref_cnt ++ would have the same effect as 'ref_cnt = 1' here.

Much of this code is unchanged otherwise.

File: prov/gni/src/gnix_mr_notifier.c [hunk 3]
```diff
@@ -128,30 +130,38 @@ _gnix_notifier_close(struct gnix_mr_notifier *mrn)
...
-		if (close(mrn->fd) != 0) {
-			ret_errno = errno;
-			GNIX_INFO(FI_LOG_MR, "error closing kdreg device: %s\n",
-				  strerror(ret_errno));
-			// Not all of these map to fi_errno values
-			ret = -ret_errno;
-			goto err_exit;
-		}
+	if (--mrn->ref_cnt) {
+		goto err_exit;
```
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c:142]: I think this is best off with being a gnix reference object and controlled via the destruct. Given that the last reference to the notifier would be dropped when the last domain died, there isn't really a chance of a race with another open.

File: prov/gni/test/mr_notifier.c [hunk 4]
```diff
@@ -43,9 +43,6 @@ static void mr_notifier_setup(void)
 {
 	int ret;
 
-	ret = _gnix_notifier_init(&mr_notifier);
-	cr_assert(ret == 0, "_gnix_notifier_init failed");
```
Comment (jswaro) [inline prov/gni/test/mr_notifier.c:47]: Why was this removed?

File: prov/gni/src/gnix_mr_notifier.c (hunk 1 above)
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c:64]: reference count should be initialized here. It then simplifies the logic at the tail of the *_open() function.

File: prov/gni/src/gnix_mr_notifier.c (hunk 2 above)
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c]: I'm recommended that you only add logic that differs from the current flow. Doing so would remove all of the spurious whitespace changes.

```C
if (already_initialized) goto increment_reference;

increment_reference:
    ref_cnt++;
```

File: prov/gni/src/gnix_mr_notifier.c [hunk 5]
```diff
@@ -55,69 +57,69 @@ notifier_verify_stuff(struct gnix_mr_notifier *mrn) {
...
 
-	if (get_user_delta_args.user_delta == NULL) {
-		GNIX_INFO(FI_LOG_MR, "kdreg get_user_delta is NULL\n");
-		ret = -FI_ENODATA;
-		goto err_exit;
+		global_mr_not.fd = kdreg_fd;
+		global_mr_not.cntr = (kdreg_user_delta_t *)
+				get_user_delta_args.user_delta;
+		global_mr_not.ref_cnt = 1;
+	} else {
```
Comment (jswaro) [inline prov/gni/src/gnix_mr_notifier.c]: if the ref_cnt was initialized during init, then both paths could simply do ref_cnt++;

**Review by jswaro — CHANGES_REQUESTED** [review]
Please see comments. There are a few things that could be cleaned up and reduce the size of the change set.

File: prov/gni/src/gnix_mr_notifier.c (hunk 1 above)
Comment (ztiffany) [reply prov/gni/src/gnix_mr_notifier.c:64]: I don't think it was necessary for the ref count to be set here.

This function is now used to initialize the library, not each fabric.  It would only be run once.

File: prov/gni/test/mr_notifier.c (hunk 4 above)
Comment (ztiffany) [reply prov/gni/test/mr_notifier.c:47]: Because it's always done in the library constructor now.

**Comment by ztiffany** [comment]
I streamlined the use of ref_cnt as you liked.  I don't see the benefit of rewriting my patch to preserve the original flow.  I don't see the benefit of using the ref count object since there is a single point here where we are doing counting.

File: prov/gni/src/gnix_mr_notifier.c (hunk 1 above)
Comment (jswaro) [reply prov/gni/src/gnix_mr_notifier.c:64]: ok

File: prov/gni/test/mr_notifier.c (hunk 4 above)
Comment (jswaro) [reply prov/gni/test/mr_notifier.c:47]: ok

**Comment by jswaro** [comment]
> I streamlined the use of ref_cnt as you liked. 

Thanks.

> I don't see the benefit of rewriting my patch to preserve the original flow.

The patch creates unnecessary line differences where there are no actual changes in behavior, only white space. This is a pain later on when we need to 'git blame' and find out what commit changed that section of code last. 

> I don't see the benefit of using the ref count object since there is a single point here where we are doing counting.

I think this is fine, given it is such a simple case. I recommended the ref object because that is how we do it everywhere else when we handle reference counting.

**Comment by jswaro** [comment]
It's fine how it is. I'd still like to see the white space changes removed but approving it regardless. 👍

**Review by jswaro — APPROVED** [review]

**Comment by sungeunchoi** [comment]
👍

File: prov/gni/test/mr_notifier.c (hunk 4 above)
Comment (sungeunchoi) [reply prov/gni/test/mr_notifier.c:47]: This is still required, because we don't do library initialization in this test.
